File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2025/Y25D05.kt
package org.gristle.pdxpuzzles.everybodycodes.y2025

import org.gristle.pdxpuzzles.everybodycodes.utilities.Day
import org.gristle.pdxpuzzles.utilities.iteration.minMax
import org.gristle.pdxpuzzles.utilities.parsing.getLongList
import kotlin.math.sign

object Y25D05 : Day {
    fun concat(a: Long, b: Long): Long {
        var pow = 1
        while (pow <= b) {
            pow *= 10
        }
        return a * pow + b
    }

    data class Segment(var spine: Long, var left: Long? = null, var right: Long? = null) {
        fun place(n: Long): Boolean =
            if (n < spine && left == null) {
                left = n
                true
            } else if (n > spine && right == null) {
                right = n
                true
            } else {
                false
            }

        fun number(): Long = concat(
            concat(left ?: 0L, spine),
            right ?: 0
        )
    }

    data class Sword(val id: Long, val segments: List<Segment>): Comparable<Sword> {
        companion object {
            fun from(s: String): Sword {
                val ints = s.getLongList()
                val id = ints[0]
                val segments = mutableListOf(Segment(ints[1]))

                outer@for (int in ints.drop(2)) {
                    for (idx in segments.indices) {
                        val segment = segments[idx]
                        if (segment.place(int)) continue@outer
                    }
                    segments.add(Segment(int))
                }

                return Sword(id, segments)
            }
        }

        fun quality(): Long = segments.fold(0L) { acc, segment ->
            concat(acc, segment.spine)
        }

        override fun compareTo(other: Sword): Int {
            val qualityControl = quality() - other.quality()
            return if (qualityControl != 0L) {
                qualityControl.sign
            } else {
                val segmentControl = segments
                    .zip(other.segments)
                    .map { (a, b) -> a.number() - b.number() }
                    .firstOrNull { it != 0L }
                segmentControl?.sign ?: (id - other.id).sign
            }
        }
    }

    override fun part1(input: String) = Sword.from(input).quality()

    override fun part2(input: String) = input
        .lines()
        .map { Sword.from(it).quality() }
        .minMax()
        .let { (min, max) -> max - min }

    override fun part3(input: String) = let {
        val swords = input
            .lines()
            .map(Sword::from)
        val sortedSwords = swords
            .sortedDescending()
        sortedSwords
            .mapIndexed { index, sword -> (index + 1) * sword.id }
            .sum()
    }
}

fun main() = Day.runDay(Y25D05::class)

//    Quest 1: 2782784532 (0ms)
//    Quest 2: 8637361015798 (7ms)
//    Quest 3: 31574813 (9ms)
//    Total time: 18ms

File: rust/everybody-codes/src/bin/2025/y25d05.rs
use everybody_codes::utilities::inputs::get_event_inputs;
use itertools::{Itertools, MinMaxResult};
use utilities::{
    parsing::get_numbers::ContainsNumbers,
    structs::stopwatch::{ReportDuration, Stopwatch},
};

type Input<'a> = &'a str;
type Int = u64;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_event_inputs(25, 5);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

struct Segment {
    spine: Int,
    left: Option<Int>,
    right: Option<Int>,
}

impl Segment {
    fn place(&mut self, int: Int) -> bool {
        if int < self.spine && self.left.is_none() {
            self.left = Some(int);
            true
        } else if int > self.spine && self.right.is_none() {
            self.right = Some(int);
            true
        } else {
            false
        }
    }

    fn number(&self) -> Int {
        concat(
            concat(self.left.unwrap_or_default(), self.spine),
            self.right.unwrap_or_default(),
        )
    }
}

struct Sword {
    id: Int,
    segments: Vec<Segment>,
}

impl Sword {
    fn quality(&self) -> Int {
        self.segments
            .iter()
            .fold(0, |acc, segment| concat(acc, segment.spine))
    }
}

impl From<&str> for Sword {
    fn from(value: &str) -> Self {
        let mut ints = value.get_numbers::<Int>();
        let id = ints.next().unwrap();
        let mut segments: Vec<Segment> = Vec::new();
        for int in ints {
            if !segments.iter_mut().any(|segment| segment.place(int)) {
                segments.push(Segment { spine: int, left: None, right: None, })
            }
        }
        Self { id, segments }
    }
}

fn concat(a: Int, b: Int) -> Int {
    let mut pow = 1;
    while pow <= b {
        pow *= 10;
    }
    a * pow + b
}

fn part1(input: Input) -> Int {
    Sword::from(input).quality()
}

fn part2(input: Input) -> Int {
    let minmax = input
        .lines()
        .map(|line| Sword::from(line).quality())
        .minmax();
    if let MinMaxResult::MinMax(min, max) = minmax {
        return max - min;
    }
    unreachable!()
}

fn part3(input: Input) -> Int {
    input
        .lines()
        .map(|line| Sword::from(line))
        .sorted_by_cached_key(|sword| {
            (
                sword.quality(),
                sword.segments.iter().map(|seg| seg.number()).collect_vec(),
                sword.id,
            )
        })
        .rev()
        .enumerate()
        .map(|(idx, sword)| (idx as Int + 1) * sword.id)
        .sum()
}

#[test]
fn default() {
    let (input1, input2, input3) = get_event_inputs(25, 5);
    assert_eq!(2782784532, part1(&input1));
    assert_eq!(8637361015798, part2(&input2));
    assert_eq!(31574813, part3(&input3));
}

// Input parsed (41μs)
// 1. 2782784532 (9μs)
// 2. 8637361015798 (107μs)
// 3. 31574813 (700μs)
// Total: 861μs
